File: go/513.go
package main

//Definition for a binary tree node.
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func findBottomLeftValue(root *TreeNode) int {
	var ret int = 0
	var q []*TreeNode = make([]*TreeNode, 0)
	var size int
	q = append(q, root)

	// a BFS method
	for len(q) > 0 {
	  // get the leftest of this layer
		ret = q[0].Val
		for size = len(q) ; size > 0; size-- {
			if q[0].Left != nil {
				q = append(q, q[0].Left)
			}
			if q[0].Right != nil {
				q = append(q, q[0].Right)
			}
			q = q[1:]
		}
	}
	return ret
}

File: cpp/513.cpp
#include <iostream>
#include <queue>

using namespace std;

// Definition for a binary tree node.
struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right)
      : val(x), left(left), right(right) {}
};

class Solution {
public:
  int findBottomLeftValue(TreeNode *root) {
      queue<TreeNode *> q;
      q.push(root);
      int ret, size;

      // BFS method
      while (!q.empty()) {
          size = q.size();
          // get the leftist of this layer
          ret = q.front()->val;
          while (size) { // I found this may use for loop when I writing Go, haha
              if (q.front()->left)
                  q.push(q.front()->left);
              if (q.front()->right)
                  q.push(q.front()->right);
              q.pop();
              size--;
          }
      }

      return ret;
  }
};
